cpp/ql/src/semmle/code/cpp/Type.qll

```diff
@@ -1710,28 +1710,6 @@ class AutoType extends TemplateParameter {
   }
 }
 
-//
-// Internal implementation predicates
-//
-private predicate allSpecifiers(int i, string s) { s = rank[i](string t | specifiers(_, t) | t) }
-
-private predicate internalSpecString(Type t, string res, int i) {
-  (
-    if allSpecifiers(i, t.getASpecifier().getName())
-    then
-      exists(string spec, string rest |
-        allSpecifiers(i, spec) and
-        res = spec + " " + rest and
-        internalSpecString(t, rest, i + 1)
-      )
-    else (
-      allSpecifiers(i, _) and internalSpecString(t, res, i + 1)
-    )
-  )
-  or
-  i = count(Specifier s) + 1 and res = ""
-}
```

Comment:
Was this just ... very old code?  (we've had `concat` for a while now!)

Reply:
I didn't dig in the history - I suspect either it's very old or written by somebody who was fairly new to QL at the time.

Reply:
Hrm. I assume that I wrote this, but it doesn't look like my style. In any case, nice improvement in both readability and performance.